### DISCUSSION TITLE
Making backend functions available under a global object instead of using `webui.call("fn_name_string", ...args)`

### DISCUSSION THREAD

**Discussion by ttytm** [webui-dev/webui#263] [Ideas]
I think the following change could add a lot of convenience.

If we bind a C function like the following:
```c
webui_bind(w, "myStrFunc", my_str_func);
```

It could became available as a JS function under a global object.
 
```js
// Instead of calling it via
webui.call("myStrFunc", myArg)
// Call the function from the global object or window
webui.myStrFunc(myArg)
```

By default `webui` could be used as object name. But there should be an api that allows to customize the name of the object and to set it to an empty string, then it would just be the window object and the it would become a global function.

**Comment by hassandraga** [comment]
I really liked this idea, Thank you for pointing this out 👍
Implemented (https://github.com/webui-dev/webui/commit/87a3a4f4e6e117db60d289350746c5a16dd55d50).

**Reply by ttytm** [reply]
Exceptional @hassandraga! I think this is a huge improvement. It makes things generally more explicit and usage more native. And when working with typescript, this allows to generate type delcarations for functions to enable assistance from language servers. 

Things like this would now work for webui functions (simplified example):

```typescript
// WebUI functions
declare function myStrFunc(str: string): Promise<string>;
...
```

In the long term we might can work out a strategy to automatically generate those definitions for bind callbacks.

Thanks a lot for the implementation!

**Reply by hassandraga** [reply]
This is beautiful!
Now from JS:
```js
my_backend_func(arg1, arg2).then((response) => {
    // ...
});
```

Implemented (https://github.com/webui-dev/webui/commit/46c5125e0d53bb7fee257644398427bc9211638d).

The only issue with this mechanism is it may cause conflict between the back-end function names and global names. To solve this, WebUI Bridge will check if a global object already exists before adding it... However, `webui.my_backend_func()` and `webui.call(my_backend_func)` still work anyway.

**Reply by ttytm** [reply]
Having it available via `webui.` is a good default I think :slightly_smiling_face: :+1: .

What do you think about giving the power to the user to set the object name?

Then something like `webui_set_callback_object` could take a char that would be used as object name.
So when setting `webui_set_callback_object("ui")` it would become `ui.my_backend_func`. An empty string `webui_set_callback_object("")` would cause things to become avilable via the global scope `my_backend_func`.

**Reply by hassandraga** [reply]
I don't think `webui_set_callback_object` is necessary. The user can easily set in JS `const ui = webui;` if needed.

**Reply by hassandraga** [reply]
However, those are always available in JS:
```js
webui.my_backend_func();
webui.call(my_backend_func);
```

This available if `my_backend_func` object does not already exist globally in JS:
```js
my_backend_func();
```

**Reply by ttytm** [reply]
Very good. Thank you!
